WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/BlockProcessor.java

```diff
@@ -30,6 +30,7 @@ public abstract class BlockProcessor {
     String mLocalId;
     String mRemoteId;
     String mRemoteUrl;
+    String mRemoteImageUrl;
```

Comment:
Why do we introduce another remote URL here? For blocks other than image (or gallery), `getOptimalFileURL` will return the same string as `getFileURL`. Could we just set `mRemoteURL` to the escaped `mediaFile.getOptimalFileURL`? Wdyt?

Reply:
I made this change after introducing `getOptimalFileURL`, around 38 tests related to media started failing because of it, so I thought of just changing the needed blocks. Another option is to keep `mRemoteUrl` and update all of the tests to mock `optimalFileUrl`, should we just use `mRemoteUrl` and update the tests?

Reply:
> around 38 tests related to media started failing because of it,

:sweat_smile: yikes! thats a lot of failing tests.

> should we just use mRemoteUrl and update the tests?

I still think we should, since it seems weird to have the implementation details of a subclass in the base class. I guess it is natural that we must change the mocks, since we are relying on this new method internally in the object / class under test. But it is rather tedious, so I've made the following patch to "chip in" :smile: :+1: 

<details>
<summary>Patch to update processor tests</summary>

```patch
diff --git a/WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/BlockProcessor.java b/WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/BlockProcessor.java
index 6f5dc0bc4d..21dba8597c 100644
--- a/WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/BlockProcessor.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/BlockProcessor.java
@@ -30,7 +30,6 @@ public abstract class BlockProcessor {
     String mLocalId;
     String mRemoteId;
     String mRemoteUrl;
-    String mRemoteImageUrl;
 
     private String mBlockName;
     private JsonObject mJsonAttributes;
@@ -45,9 +44,8 @@ public abstract class BlockProcessor {
     BlockProcessor(String localId, MediaFile mediaFile) {
         mLocalId = localId;
         mRemoteId = mediaFile.getMediaId();
-        mRemoteUrl = org.wordpress.android.util.StringUtils.notNullStr(Utils.escapeQuotes(mediaFile.getFileURL()));
-        mRemoteImageUrl = org.wordpress.android.util.StringUtils.notNullStr(Utils
-                .escapeQuotes(mediaFile.getOptimalFileURL()));
+        mRemoteUrl = org.wordpress.android.util.StringUtils.notNullStr(Utils.escapeQuotes(mediaFile
+                .getOptimalFileURL()));
     }
 
     private JsonObject parseJson(String blockJson) {
diff --git a/WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/GalleryBlockProcessor.java b/WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/GalleryBlockProcessor.java
index a44ffe8fd3..ebfaea3ced 100644
--- a/WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/GalleryBlockProcessor.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/GalleryBlockProcessor.java
@@ -49,7 +49,7 @@ public class GalleryBlockProcessor extends BlockProcessor {
         // if a match is found, proceed with replacement
         if (targetImg != null) {
             // replace attributes
-            targetImg.attr("src", mRemoteImageUrl);
+            targetImg.attr("src", mRemoteUrl);
             targetImg.attr("data-id", mRemoteId);
             targetImg.attr("data-full-url", mRemoteUrl);
             targetImg.attr("data-link", mAttachmentPageUrl);
@@ -63,7 +63,7 @@ public class GalleryBlockProcessor extends BlockProcessor {
             if (parent != null && parent.is("a") && mLinkTo != null) {
                 switch (mLinkTo) {
                     case "file":
-                        parent.attr("href", mRemoteImageUrl);
+                        parent.attr("href", mRemoteUrl);
                         break;
                     case "post":
                         parent.attr("href", mAttachmentPageUrl);
diff --git a/WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/ImageBlockProcessor.java b/WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/ImageBlockProcessor.java
index 48864e5c4e..16c26897c6 100644
--- a/WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/ImageBlockProcessor.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/ImageBlockProcessor.java
@@ -19,7 +19,7 @@ public class ImageBlockProcessor extends BlockProcessor {
         // if a match is found, proceed with replacement
         if (targetImg != null) {
             // replace attributes
-            targetImg.attr("src", mRemoteImageUrl);
+            targetImg.attr("src", mRemoteUrl);
 
             // replace class
             targetImg.removeClass("wp-image-" + mLocalId);
diff --git a/WordPress/src/test/java/org/wordpress/android/ui/posts/PostUtilsUploadProcessingTest.kt b/WordPress/src/test/java/org/wordpress/android/ui/posts/PostUtilsUploadProcessingTest.kt
index d5346e50dd..7d1086a418 100644
--- a/WordPress/src/test/java/org/wordpress/android/ui/posts/PostUtilsUploadProcessingTest.kt
+++ b/WordPress/src/test/java/org/wordpress/android/ui/posts/PostUtilsUploadProcessingTest.kt
@@ -19,7 +19,7 @@ class PostUtilsUploadProcessingTest {
     @Before
     fun before() {
         whenever(mediaFile.mediaId).thenReturn(TestContent.remoteMediaId)
-        whenever(mediaFile.fileURL).thenReturn(TestContent.remoteImageUrl)
+        whenever(mediaFile.optimalFileURL).thenReturn(TestContent.remoteImageUrl)
         whenever(mediaFile.optimalFileURL).thenReturn(TestContent.remoteImageUrl)
         whenever(mediaFile.getAttachmentPageURL(any())).thenReturn(TestContent.attachmentPageUrl)
     }
@@ -49,7 +49,7 @@ class PostUtilsUploadProcessingTest {
 
     @Test
     fun `replaceMediaFileWithUrlInGutenbergPost also works with video`() {
-        whenever(mediaFile.fileURL).thenReturn(TestContent.remoteVideoUrl)
+        whenever(mediaFile.optimalFileURL).thenReturn(TestContent.remoteVideoUrl)
         val processedContent = PostUtils.replaceMediaFileWithUrlInGutenbergPost(TestContent.oldMediaTextBlockWithVideo,
                 TestContent.localMediaId, mediaFile, TestContent.siteUrl)
         Assertions.assertThat(processedContent).isEqualTo(TestContent.newMediaTextBlockWithVideo)
diff --git a/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/AudioBlockProcessorTest.kt b/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/AudioBlockProcessorTest.kt
index b6c4aa2322..8f89cbee52 100644
--- a/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/AudioBlockProcessorTest.kt
+++ b/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/AudioBlockProcessorTest.kt
@@ -14,7 +14,7 @@ class AudioBlockProcessorTest {
     @Before
     fun before() {
         whenever(mediaFile.mediaId).thenReturn(TestContent.remoteMediaId)
-        whenever(mediaFile.fileURL).thenReturn(TestContent.remoteAudioUrl)
+        whenever(mediaFile.optimalFileURL).thenReturn(TestContent.remoteAudioUrl)
         processor = AudioBlockProcessor(TestContent.localMediaId, mediaFile)
     }
 
diff --git a/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/CoverBlockProcessorTest.kt b/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/CoverBlockProcessorTest.kt
index 40dce17930..ff5f638012 100644
--- a/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/CoverBlockProcessorTest.kt
+++ b/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/CoverBlockProcessorTest.kt
@@ -17,7 +17,7 @@ class CoverBlockProcessorTest {
     @Before
     fun before() {
         whenever(mediaFile.mediaId).thenReturn(TestContent.remoteMediaId)
-        whenever(mediaFile.fileURL).thenReturn(TestContent.remoteImageUrl)
+        whenever(mediaFile.optimalFileURL).thenReturn(TestContent.remoteImageUrl)
         processor = CoverBlockProcessor(TestContent.localMediaId, mediaFile, mediaUploadCompletionProcessor)
     }
 
@@ -43,7 +43,7 @@ class CoverBlockProcessorTest {
     @Test
     fun `processBlock works with outer nested cover blocks`() {
         whenever(mediaFile.mediaId).thenReturn(TestContent.remoteMediaId2)
-        whenever(mediaFile.fileURL).thenReturn(TestContent.remoteImageUrl2)
+        whenever(mediaFile.optimalFileURL).thenReturn(TestContent.remoteImageUrl2)
         processor = CoverBlockProcessor(TestContent.localMediaId2, mediaFile, mediaUploadCompletionProcessor)
         val processedBlock = processor.processBlock(TestContent.oldCoverBlockWithNestedCoverBlockOuter)
         Assertions.assertThat(processedBlock).isEqualTo(TestContent.newCoverBlockWithNestedCoverBlockOuter)
@@ -63,7 +63,7 @@ class CoverBlockProcessorTest {
 
     @Test
     fun `processBlock works with videos`() {
-        whenever(mediaFile.fileURL).thenReturn(TestContent.remoteVideoUrl)
+        whenever(mediaFile.optimalFileURL).thenReturn(TestContent.remoteVideoUrl)
         processor = CoverBlockProcessor(TestContent.localMediaId, mediaFile, mediaUploadCompletionProcessor)
         val processedBlock = processor.processBlock(TestContent.oldCoverBlockWithVideo)
         Assertions.assertThat(processedBlock).isEqualTo(TestContent.newCoverBlockWithVideo)
diff --git a/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/FileBlockProcessorTest.kt b/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/FileBlockProcessorTest.kt
index 20c5f00332..5e217db6be 100644
--- a/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/FileBlockProcessorTest.kt
+++ b/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/FileBlockProcessorTest.kt
@@ -14,7 +14,7 @@ class FileBlockProcessorTest {
     @Before
     fun before() {
         whenever(mediaFile.mediaId).thenReturn(TestContent.remoteMediaId)
-        whenever(mediaFile.fileURL).thenReturn(TestContent.remoteImageUrl)
+        whenever(mediaFile.optimalFileURL).thenReturn(TestContent.remoteImageUrl)
         processor = FileBlockProcessor(TestContent.localMediaId, mediaFile)
     }
 
diff --git a/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/MediaTextBlockProcessorTest.kt b/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/MediaTextBlockProcessorTest.kt
index 06a5c2b1b7..75302ac8d6 100644
--- a/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/MediaTextBlockProcessorTest.kt
+++ b/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/MediaTextBlockProcessorTest.kt
@@ -15,7 +15,7 @@ class MediaTextBlockProcessorTest {
     @Before
     fun before() {
         whenever(mediaFile.mediaId).thenReturn(TestContent.remoteMediaId)
-        whenever(mediaFile.fileURL).thenReturn(TestContent.remoteImageUrl)
+        whenever(mediaFile.optimalFileURL).thenReturn(TestContent.remoteImageUrl)
         processor = MediaTextBlockProcessor(TestContent.localMediaId, mediaFile)
     }
 
@@ -33,7 +33,7 @@ class MediaTextBlockProcessorTest {
 
     @Test
     fun `processBlock also works for video`() {
-        whenever(mediaFile.fileURL).thenReturn(TestContent.remoteVideoUrl)
+        whenever(mediaFile.optimalFileURL).thenReturn(TestContent.remoteVideoUrl)
         processor = MediaTextBlockProcessor(TestContent.localMediaId, mediaFile)
         val processedBlock = processor.processBlock(TestContent.oldMediaTextBlockWithVideo)
         Assertions.assertThat(processedBlock).isEqualTo(TestContent.newMediaTextBlockWithVideo)
diff --git a/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/MediaUploadCompletionProcessorTest.kt b/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/MediaUploadCompletionProcessorTest.kt
index 09c42f918b..d68903b8fb 100644
--- a/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/MediaUploadCompletionProcessorTest.kt
+++ b/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/MediaUploadCompletionProcessorTest.kt
@@ -19,7 +19,7 @@ class MediaUploadCompletionProcessorTest {
     @Before
     fun before() {
         whenever(mediaFile.mediaId).thenReturn(TestContent.remoteMediaId)
-        whenever(mediaFile.fileURL).thenReturn(TestContent.remoteImageUrl)
+        whenever(mediaFile.optimalFileURL).thenReturn(TestContent.remoteImageUrl)
         whenever(mediaFile.optimalFileURL).thenReturn(TestContent.remoteImageUrl)
         whenever(mediaFile.getAttachmentPageURL(any())).thenReturn(TestContent.attachmentPageUrl)
         processor = MediaUploadCompletionProcessor(TestContent.localMediaId, mediaFile, TestContent.siteUrl)
@@ -33,7 +33,7 @@ class MediaUploadCompletionProcessorTest {
 
     @Test
     fun `processPost splices id and url for a video block`() {
-        whenever(mediaFile.fileURL).thenReturn(TestContent.remoteVideoUrl)
+        whenever(mediaFile.optimalFileURL).thenReturn(TestContent.remoteVideoUrl)
         processor = MediaUploadCompletionProcessor(TestContent.localMediaId, mediaFile, TestContent.siteUrl)
         val blocks = processor.processContent(TestContent.oldPostVideo)
         Assertions.assertThat(blocks).isEqualTo(TestContent.newPostVideo)
@@ -66,7 +66,7 @@ class MediaUploadCompletionProcessorTest {
     @Test
     fun `processPost works for nested outer cover blocks`() {
         whenever(mediaFile.mediaId).thenReturn(TestContent.remoteMediaId2)
-        whenever(mediaFile.fileURL).thenReturn(TestContent.remoteImageUrl2)
+        whenever(mediaFile.optimalFileURL).thenReturn(TestContent.remoteImageUrl2)
         processor = MediaUploadCompletionProcessor(TestContent.localMediaId2, mediaFile, TestContent.siteUrl)
         val blocks = processor.processContent(TestContent.oldCoverBlockWithNestedCoverBlockOuter)
         Assertions.assertThat(blocks).isEqualTo(TestContent.newCoverBlockWithNestedCoverBlockOuter)
diff --git a/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/VideoBlockProcessorTest.kt b/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/VideoBlockProcessorTest.kt
index b548503a5b..355df30e3c 100644
--- a/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/VideoBlockProcessorTest.kt
+++ b/WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/VideoBlockProcessorTest.kt
@@ -15,7 +15,7 @@ class VideoBlockProcessorTest {
     @Before
     fun before() {
         whenever(mediaFile.mediaId).thenReturn(TestContent.remoteMediaId)
-        whenever(mediaFile.fileURL).thenReturn(TestContent.remoteVideoUrl)
+        whenever(mediaFile.optimalFileURL).thenReturn(TestContent.remoteVideoUrl)
         processor = VideoBlockProcessor(TestContent.localMediaId, mediaFile)
     } 
```
</details>

We only "save" two lines of code with all of this, but I think it's also _cleaner_ to not "pollute" the base class this way. Wdyt?

Reply:
Thank you so much for providing the patch! I've applied it in https://github.com/wordpress-mobile/WordPress-Android/pull/16875/commits/12ef532223a7ceb3dfe40747bdb48565ccbc329a